### PR TITLE
use bytes

### PR DESCRIPTION
**Pull request by ctaggart**
@arnodb, I reviewed https://github.com/Azure/azure-sdk-for-rust/pull/284. Let me know what you think of these changes.

### CODE REVIEW

**Review by arnodb — COMMENTED** [review]
The general idea is good, I should have done that myself.

But one thing is probably not so good: using `Bytes`. The length of the buffers are never supposed to vary at runtime, nor to vary at compile time.

Using `Bytes` has bad consequences:

- The information about the const length is lost when the consumer needs to convert a digest to, e.g. a `[u8; 8/16]`, which means the consumer will need to check the length once again.
- Bytes are less easily manipulable than `[u8]`. In my case, instead of converting them back to a slice, I decided to convert my `Digest` to a `Bytes` and that requires a copy of the data, as far as I understand.

File: sdk/storage/src/core/mod.rs [hunk 1]
```diff
@@ -42,77 +43,83 @@ use serde::{Deserialize, Deserializer};
...
-                if content_md5_vec.len() != 16 {
-                    return Err(serde::de::Error::custom(
-                        crate::Error::DigestNot16BytesLong(content_md5_vec.len() as u64),
-                    ));
-                }
-                let mut content_md5 = [0; 16];
-                content_md5.copy_from_slice(&content_md5_vec[0..16]);
-                Ok(Self(content_md5))
-            })
+        let bytes = <&[u8]>::deserialize(deserializer)?;
```
Comment (arnodb) [inline sdk/storage/src/core/mod.rs]: This breaks XML deserialization: `expected a borrowed byte array`.

File: sdk/storage/src/core/mod.rs [hunk 2]
```diff
@@ -42,77 +43,83 @@ use serde::{Deserialize, Deserializer};
...
-                if content_crc64_vec.len() != 8 {
-                    return Err(serde::de::Error::custom(crate::Error::CRC64Not8BytesLong(
-                        content_crc64_vec.len() as u64,
-                    )));
-                }
-                let mut content_crc64 = [0; 8];
-                content_crc64.copy_from_slice(&content_crc64_vec[0..8]);
-                Ok(Self(content_crc64))
-            })
+        let bytes = <&[u8]>::deserialize(deserializer)?;
```
Comment (arnodb) [inline sdk/storage/src/core/mod.rs]: This breaks XML deserialization: `expected a borrowed byte array`.

**Review by arnodb — CHANGES_REQUESTED** [review]

**Comment by arnodb** [comment]
@ctaggart I added 3 integration tests to the original branch so that we are sure we do not regress. If you rebase then you will see where the deserialization is broken.

File: sdk/storage/src/core/mod.rs (hunk 2 above)
Comment (ctaggart) [reply sdk/storage/src/core/mod.rs]: It does. Thanks for the new tests! I changed it back to StringDeserializer like you had.

**Comment by ctaggart** [comment]
> The information about the const length is lost when the consumer needs to convert a digest to, e.g. a [u8; 8/16], which means the consumer will need to check the length once again.

I added an `as_slice` and implemented AsRef. We could also do a `From` in a similar way if needed.

> Bytes are less easily manipulable than [u8]. In my case, instead of converting them back to a slice, I decided to convert my Digest to a Bytes and that requires a copy of the data, as far as I understand.

No copy, [Bytes::from](https://docs.rs/bytes/1.0.1/src/bytes/bytes.rs.html#798-828) looks to be taking ownership. Also, FYI, `Bytes` is extremely cheap to clone if you want to you it. See https://docs.rs/bytes/1.0.1/bytes/struct.Bytes.html.

**Comment by arnodb** [comment]
With `as_slice` it seems acceptable.

**Pull request merged**